src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperTableValue.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Core.IR;
+
+namespace Microsoft.PowerFx.Types
+{
+    // Project the correct compile-time type onto the runtime value. 
+    // Important for union / intersection types, such as Table() or If(). For example:
+    //    First(Table({a:1},{b:2})), result is a record with both fields a and b. 
+    internal class CompileTimeTypeWrapperTableValue : CollectionTableValue<DValue<RecordValue>>
+    {
+        private readonly TableValue _inner;
+
```

Comment:
actually, I wonder if we can remove this field altogether?

You're using inheritence to forward, which looks like it solves this. 
In fact, maybe this means we  can even remove it from CompileTimeTypeWrapperRecordValue. 

This is the classic "ineheritence vs. encapsulation" design pattern debate. :)